feat: add following filter for public, publicSummary, and publicUpdates methods

achou11 opened the pull request:
This PR adds an option called `following` to the `public`, `publicSummary`, and `publicUpdates` methods that filters the information returned to threads authored by someone the user directly follows.

TBH incredibly shaky on some of the logic and wouldn't be surprised if follow ups are needed to fix a bug I haven't caught. Consider this a first reviewable draft 😄 Also had to update various tests to add a wait period (i.e. changing `wait(cb)` to `wait(cb, 800)`), which has me a little worried (maybe that's okay?).

staltz commented on src/index.ts:
```diff
@@ -119,17 +119,67 @@ function makePassesFilter(opts: FilterOpts): (msg: Msg) => boolean {
+
+          const rootMsgKey = getRootMsgId(msg);
+
+          if (rootMsgKey === msg.key) {
```
I suspect this will be true always, in `public()` and `publicSummary()` (let's talk about `publicUpdates()` later since it's a special case and maybe should be implemented separately). In fact maybe you can `test.skip` the publicUpdates and run `npm run coverage`, I think the coverage results would show that the branch on line 166 is never run.

This is good news because it means you can delete code. i.e. lines 166–178 and you can also assume that line 158 is always true.

staltz commented on src/index.ts:
```diff
@@ -119,17 +119,67 @@ function makePassesFilter(opts: FilterOpts): (msg: Msg) => boolean {
+      pull(
+        source,
+        pull.asyncMap((msg: Msg, cb: CB<Msg | null>) => {
+          if (!enabled) {
```
We can move this check up, before `pull()` on line 137 is called. It could be `enabled ? pull.through() : pull( source, ...`

achou11 replied:
ah `pull.through` was what I was looking for. would it make sense to avoid the `enabled` arg here altogether and handle this one level up in the public methods?

staltz commented on src/index.ts:
```diff
@@ -119,17 +119,67 @@ function makePassesFilter(opts: FilterOpts): (msg: Msg) => boolean {
+            return;
+          }
+
+          if (!this.isFollowing) {
```
Maybe we can remove this check, and instead put it early inside `public()` and `publicSummary()` as an error. I'll put a comment down there suggesting how.

achou11 replied:
good suggestion - only thing is that TS will complain when this is removed so it'd be needed in one form or another within this function

staltz commented on src/index.ts:
```diff
@@ -275,6 +325,7 @@ class threads {
     const threadMaxSize = opts.threadMaxSize ?? Infinity;
+    const followingOnly = opts.following ?? false;
     const filterOperator = makeFilterOperator(opts);
     const passesFilter = makePassesFilter(opts);
```
We could add

```js
if (followingOnly && !this.isFollowing) {
  return pull.error(new Error('ssb-threads requires ssb-friends installed'))
}
```

Also for `publicSummary`

staltz commented on test/index.js:
```diff
@@ -1832,3 +1843,277 @@ test('threads.threadUpdates can view private replies given opts.private', (t) =>
   );
 });
+
+test('threads.public respects following opt', (t) => {
```
The tests for threads.public and threads.publicSummary look good!

staltz commented on src/index.ts:
```diff
@@ -119,17 +119,67 @@ function makePassesFilter(opts: FilterOpts): (msg: Msg) => boolean {
+          const rootMsgKey = getRootMsgId(msg);
+
+          if (rootMsgKey === msg.key) {
+            this.isFollowing(
```
We have to remember that the current user is also supposed to have their root posts show on the feed, so that means that early in this function we can check whether `msg.value.author === this.ssb.id` and allow that msg to pass, because I believe `isFollowing` would return false.

achou11 replied:
Does this include contact messages? After some refactoring, noticing that my `threads.public` test fails with the operations (from Lucy's db):

1. Lucy follows Mary
2. Mary creates root post
3. Alice creates root post

The test is expecting just Mary's post to be returned by the method, but it's also returning a thread with the contact message in 1. Is that the expected behavior?

achou11 replied:
Made the assumption that ^ is the expectation, so updated the test to account for contact messages

staltz replied:
That was the correct assumption, because publicUpdates in the test is considering all the message types. Typically we pass a "filter" to the publicUpdates where we tell it we are only interested in (say) `post` messages, such that the `contact` message would be left out.

staltz reviewed (changes requested):
No pressure to get it right all at once! This is tricky indeed, but I think you did a great job (the fact that the tests pass is great). I have a couple of suggestions to make, but I haven't yet given thought to `publicUpdates`, I'll have to come back to this one before we merge.

staltz commented on readme.md:
```diff
@@ -63,6 +63,7 @@ Returns a pull stream that emits thread objects of public messages.
 * `opts.threadMaxSize`: optional number (default: Infinity). Dictates the maximum amount of messages in each returned thread object. Serves for previewing threads, particularly long ones.
 * `opts.allowlist`: optional array of strings. Dictates which messages **types** to allow as root messages, while forbidding other types.
 * `opts.blocklist`: optional array of strings. Dictates which messages **types** to forbid as root messages, while allowing other types.
+- `opts.following`: optional boolean (default: false). `true` means only threads created by those directly followed will be emitted. Requires `ssb-friends`.
```
Perhaps we should add "only threads created by those directly followed (and yourself) will be emitted" just to clarify

achou11 commented:
@staltz see https://github.com/ssbc/ssb-threads/pull/18/commits/c112d7a6b51bb5dba177248a4a430c0c46ed1075 for refactor based on your feedback. Most notably created a separate filter for `publicUpdates` and simplified the one used for `public` and `publicSummary`. Also updated tests based on my assumption for https://github.com/ssbc/ssb-threads/pull/18#discussion_r980240729 and just to be clearer and more resilient to changes

staltz commented:
@achou11 Oh, great! I didn't expect to get an update here until Friday, but I can review your updates today. :)

staltz commented on src/index.ts:
```diff
@@ -115,21 +115,85 @@ function makePassesFilter(opts: FilterOpts): (msg: Msg) => boolean {
 
+function assertFollowingOnlyUsability<
+  T extends ((obj: any, cb: CB<boolean>) => void) | null,
+>(enabled: boolean, isFollowing: T): asserts isFollowing is NonNullable<T> {
```
Cool, I did not know about the existence of `asserts ... is NonNullable<T>` in TypeScript.

achou11 replied:
yeah basically an extension of the `is` keyword and is useful for cases where the asserted variable is used way later in the scope of the calling function. I've used it sparingly but handy to know of

staltz commented:
I checked the code carefully and it seems all correct. Tests passing, that's great. Thank you @achou11 !

staltz commented:
`ssb-threads@10.1.0` released
